In [1]:
# 
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error
from matplotlib import pyplot as plt

In [2]:
geo_data_0 = pd.read_csv('./geo_data_0.csv')
geo_data_1 = pd.read_csv('./geo_data_1.csv')
geo_data_2 = pd.read_csv('./geo_data_2.csv')

In [3]:
geo_data_0.info()
geo_data_1.info()
geo_data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null 

In [4]:
geo_data_0_target = geo_data_0['product']
geo_data_0_features = geo_data_0
geo_data_0_features_train, geo_data_0_features_test, geo_data_0_target_train, geo_data_0_target_test = train_test_split(geo_data_0_features, geo_data_0_target, train_size=0.75, random_state=12345)
display(geo_data_0_features_train.shape, geo_data_0_target_test.shape)

(75000, 5)

(25000,)

In [5]:
model_0 = LinearRegression()
model_0.fit(geo_data_0_features_train[['f0', 'f1', 'f2']], geo_data_0_target_train)
geo_data_0_features_test['predict'] = model_0.predict(geo_data_0_features_test[['f0', 'f1', 'f2']])

In [6]:
display(geo_data_0_features_test['predict'].mean(), root_mean_squared_error(geo_data_0_target_test, geo_data_0_features_test['predict']))

92.59256778438035

37.5794217150813

Temos a media das predicoes em 92.59 e o REQM 37.58, logo observamos uma longa distancia media, cerca de 1/3 distante da media real, o que e uma previsao muito ruim.

In [7]:
geo_data_1_target = geo_data_1['product']
geo_data_1_features = geo_data_1
geo_data_1_features_train, geo_data_1_features_test, geo_data_1_target_train, geo_data_1_target_test = train_test_split(geo_data_1_features, geo_data_1_target, train_size=0.75, random_state=12345)
display(geo_data_1_features_train.shape, geo_data_1_target_test.shape)

(75000, 5)

(25000,)

In [8]:
model_1 = LinearRegression()
model_1.fit(geo_data_1_features_train[['f0', 'f1', 'f2']], geo_data_1_target_train)
geo_data_1_features_test['predict'] = model_1.predict(geo_data_1_features_test[['f0', 'f1', 'f2']])

In [9]:
display(geo_data_1_features_test['predict'].mean(), root_mean_squared_error(geo_data_1_target_test, geo_data_1_features_test['predict']))

68.72854689544599

0.8930992867756169

Temos uma media de previsao em 68.73 e um REQM de 0.89, o que parece muito bom pra ser verdade, mas e o que observamos.

In [10]:
geo_data_2_target = geo_data_2['product']
geo_data_2_features = geo_data_2
geo_data_2_features_train, geo_data_2_features_test, geo_data_2_target_train, geo_data_2_target_test = train_test_split(geo_data_2_features, geo_data_2_target, train_size=0.75, random_state=12345)
display(geo_data_2_features_train.shape, geo_data_2_target_test.shape)

(75000, 5)

(25000,)

In [11]:
model_2 = LinearRegression()
model_2.fit(geo_data_2_features_train[['f0', 'f1', 'f2']], geo_data_2_target_train)
geo_data_2_features_test['predict'] = model_2.predict(geo_data_2_features_test[['f0', 'f1', 'f2']])

In [12]:
display(geo_data_2_features_test['predict'].mean(), root_mean_squared_error(geo_data_2_target_test, geo_data_2_features_test['predict']))

94.96504596800489

40.02970873393434

Temos a media das predicoes em 94.96 e o REQM 40.03, logo observamos uma longa distancia media, quase 1/2 distante da media real, o que e uma previsao muito ruim.

In [13]:
geo_0_mean = geo_data_0['product'].mean()
geo_1_mean = geo_data_1['product'].mean()
geo_2_mean = geo_data_2['product'].mean()
display(geo_0_mean, geo_1_mean, geo_2_mean)
print(f"A regiao 0 produz em media {(geo_0_mean - 111.1):.2f} unidades do que o esperado.")
print(f"A regiao 1 produz em media {(geo_1_mean - 111.1):.2f} unidades do que o esperado.")
print(f"A regiao 2 produz em media {(geo_2_mean - 111.1):.2f} unidades do que o esperado.")

92.50000000000001

68.82500000000002

95.00000000000004

A regiao 0 produz em media -18.60 unidades do que o esperado.
A regiao 1 produz em media -42.27 unidades do que o esperado.
A regiao 2 produz em media -16.10 unidades do que o esperado.


Calculei as medias de cada regiao e comparei com o valor minimo necessario para dar lucro, os pocos nas 3 regioes dao em media prejuizo.

In [14]:
def profit_geo_data_top_200(geo_data):
    top_200 = geo_data.sort_values(by='predict')['product'][-200:]
    return top_200.sum() * 4500 - 100000000

print(f"${(profit_geo_data_top_200(geo_data_0_features_test)):,.2f}")
print(f"${(profit_geo_data_top_200(geo_data_1_features_test)):,.2f}")
print(f"${(profit_geo_data_top_200(geo_data_2_features_test)):,.2f}")

$33,208,260.43
$24,150,866.97
$27,103,499.64


Apos as apuracoes conclui que a localizacao 0 e mais promissora que as demais de acordo com o observado pelas 200 melhores previsoes de cada regiao.

In [15]:
def bootstrap(geo_data):
    N_ITER = 1000
    profits = []
    for n in range(N_ITER):
        subsample = geo_data.sample(n=500, replace=True)
        profits.append(profit_geo_data_top_200(subsample))
    return np.array(profits)

profits_geo = [bootstrap(geo_data) for geo_data in [geo_data_0_features_test, geo_data_1_features_test, geo_data_2_features_test]]

In [16]:
for i, profit in enumerate(profits_geo):
    mean_profit = profit.mean()
    risk = (profit < 0).mean()
    conf_interval = np.quantile(profit, [0.025, 0.975])
    print(f"Regiao {i}:\nLucro Medio: {mean_profit:,.2f}\nRisco:{(risk*100):.2f}%\nIntervalo de probabilidade de encontrar 95% dos dados:[{conf_interval[0]:,.2f}, {conf_interval[1]:,.2f}]\n")

Regiao 0:
Lucro Medio: 3,912,563.94
Risco:6.60%
Intervalo de probabilidade de encontrar 95% dos dados:[-1,211,756.16, 8,843,710.79]

Regiao 1:
Lucro Medio: 4,574,206.55
Risco:1.60%
Intervalo de probabilidade de encontrar 95% dos dados:[360,130.54, 8,687,475.30]

Regiao 2:
Lucro Medio: 3,938,971.86
Risco:6.50%
Intervalo de probabilidade de encontrar 95% dos dados:[-1,289,002.86, 9,210,669.14]



Apos analizar a os dados por regial num intervalo de confianca de 95% e calcular seu lucro medio e o risco de dar prejuizo usando bootstrap, observamos que por mais improvavel que pareceu antariormente a regiao 1 surpreendeu e apresentou os melhores resultados em todos os quesitos. A regiao 1 seria a escolhida.